## Diagonalisation

- If $ \textbf{A} $ is an $ n \times n $ matrix, then $ \textbf{A} $ is diagonalizable if and only if $ \textbf{A} $ has $ n $ linearly independent eigenvectors.

- If $ \textbf{A} $ is an $ n \times n $ matrix, with $ n $ distinct eigenvalues $ \lambda_1, \lambda_2, \dots, \lambda_n $. Then $ \textbf{A} $ is diagonalizable.

- $ \textbf{A} $ is diagonalizable if and only if the geometric multiplicity of every eigenvalue is equal to its algebraic multiplicity.


A matrix $ \textbf{A} $ is called diagonalizable (off-diagonal entries are all zero) if it is similar to a diagonal matrix $ \textbf{D} $, and there exists an invertible $ \textbf{P} $ such that $ \textbf{D} = \textbf{P}^{−1}\textbf{A}\textbf{P} $ is a diagonal matrix. Conversely, $ \textbf{A} = \textbf{P}\textbf{D}\textbf{P}^{−1} $

The diagonal matrix $ \textbf{D} $ entries are the eigenvalues of matrix $ \textbf{A} $. The columns of matrix $ \textbf{P} $ are the eigenvectors of matrix $ \textbf{A} $.

### Theorem

If $ \textbf{A} $ is an $ n \times n $ matrix, then $ \textbf{A} $ is diagonalizable if and only if $ \textbf{A} $ has n linearly independent eigenvectors.

**Proof**

Assume $ \textbf{A} $ is diagonalizable

$ \textbf{P}^{-1}\textbf{A}\textbf{P} = \textbf{D} $

$ \Rightarrow \textbf{A}\textbf{P} = \textbf{P}\textbf{D} \ \ \ \ \ \ \ [1]$

Let $ \ \textbf{p}_1, \ \textbf{p}_2, \ \dots, \ \textbf{p}_n \ $ denote the column vectors of $ \textbf{P} $.

Let $ \lambda_1, \lambda_2, \dots, \lambda_n $ denote the diagonal entries of $ \textbf{D} $. But since the diagonal elements of a diagonal matrix are also its eigenvalues, then $ \lambda_1, \lambda_2, \dots, \lambda_n $ are also the eigenvalues of $ \textbf{D} $.

By the similarity principle, the eigenvalues of $ \textbf{D} = \textbf{P}^{−1}\textbf{A}\textbf{P} $ are also the eigevalues of $ \textbf{A} $.

Now we can write equation [1] as

$ \textbf{AP} = \textbf{A}[\textbf{p}_1, \textbf{p}_2, \dots, \textbf{p}_n] $

and

$ \textbf{PD} = [\lambda_1 \textbf{p}_1, \lambda_2\textbf{p}_2, \dots, \lambda_n\textbf{p}_n] $

Thus we can write 

$ \textbf{A}\textbf{p}_1 = \lambda_1 \textbf{p}_1 $, $\ \ \textbf{A}\textbf{p}_2 = \lambda_2 \textbf{p}_2 $, $ \dots $, $  \ \textbf{A}\textbf{p}_n = \lambda_n \textbf{p}_n $

But each one of these equations is the definition of an eigenvector of $ \textbf{A} $. Therefore $ \ \textbf{p}_1, \ \textbf{p}_2, \ \dots, \ \textbf{p}_n \ $ are eigenvectors of $ \textbf{A} $.

Now since $ \textbf{P} $ is an invertible matrix (has an inverse $ \textbf{P}^{-1} $), then its columns are linearly independent (see definition of inverse matrix).

Therefore we have proved that the eigevectors $ \ \textbf{p}_1, \ \textbf{p}_2, \ \dots, \ \textbf{p}_n \ $ of $ \textbf{A} $ are linearly independent.

#### Example 1

Find a matrix $ \textbf{P} $ that diagonalizes the matrix $ \textbf{A} $

$ 
\textbf{A} = 
\begin{bmatrix}
-1 & 1 & 0 \\
0 & 3 & 0 \\
4 & -2 & 5
\end{bmatrix}
$

**Solution:**

We need to find $ \textbf{P} $ such that $ \textbf{D} = \textbf{P}^{-1} \textbf{A} \textbf{P} $

First, find the eigenvalues of $ \textbf{A} $

In [6]:
import numpy as np
import scipy as sp

A = np.array([[-1, 1, 0], [0, 3, 0], [4, -2, 5]])

w, v= sp.linalg.eig(A)

print('List of Eigenvalues', w)
print('\nList of Eigenvectors\n', v.transpose())

List of Eigenvalues [ 5.+0.j -1.+0.j  3.+0.j]

List of Eigenvectors
 [[ 0.          0.          1.        ]
 [ 0.83205029  0.         -0.5547002 ]
 [ 0.21821789  0.87287156  0.43643578]]


Therefore the eigenvalues and eigenvectors (normalised) of $ \textbf{A} $ are 

$ \lambda_1 = 5 $ with eigenvector 
$ 
\begin{bmatrix}
0 \\
0 \\
1
\end{bmatrix}
$

$ \lambda_2 = -1 $ with eigenvector
$ 
\begin{bmatrix}
0.832 \\
0 \\
-0.555
\end{bmatrix}
$


$ \lambda_3 = 3 $ with eigenvector
$ 
\begin{bmatrix}
0.218 \\
0.873 \\
0.436
\end{bmatrix}
$

Now by definition, the diagonlised matrix will have the eigenvalues in the diagonal elements

$ 
\textbf{D} = 
\begin{bmatrix}
5 & 0 & 0 \\
0 & -1 & 0 \\
0 & 0 & 3
\end{bmatrix}
$

such that $ \textbf{D} = \textbf{P}^{-1} \textbf{A} \textbf{P} $

The above eigenvectors make up the columns of the $ \textbf{P} $ matrix

$ \Rightarrow 
\textbf{P} = 
\begin{bmatrix}
0 & 0.832 & 0.218 \\
0 & 0 & 0.873 \\
1 & -0.555 & 0.436
\end{bmatrix}
$

Now find the inverse of this matrix, $ \textbf{P}^{-1} $

In [7]:
import numpy as np
import scipy as sp

p = np.array([[0, 0.832, 0.218], [0, 0, 0.873], [1, -0.555, 0.436]])

p_inv = sp.linalg.inv(p)

print(p_inv)

[[ 0.66706731 -0.66600306  1.        ]
 [ 1.20192308 -0.30013658  0.        ]
 [ 0.          1.14547537  0.        ]]


Therefore

$ 
\textbf{P}^{-1} = 
\begin{bmatrix}
0.667 & -0.666 & 1 \\
1.202 & -0.300 & 0 \\
0 & 1.145 & 0
\end{bmatrix}
$

Finally, evaluate $ \textbf{P}^{-1} \textbf{A} \textbf{P} $ to find the diagnolised version of $ \textbf{A} $

In [9]:
import numpy as np
import scipy as sp

A = np.array([[-1, 1, 0], [0, 3, 0], [4, -2, 5]])

P = np.array([[0, 0.832, 0.218], [0, 0, 0.873], [1, -0.555, 0.436]])

P_inv = sp.linalg.inv(P)

print(P_inv.dot(A).dot(P))

[[ 5.00000000e+00 -2.00000000e-03 -1.33293269e-03]
 [ 0.00000000e+00 -1.00000000e+00  1.20192308e-03]
 [ 0.00000000e+00  0.00000000e+00  3.00000000e+00]]


Which is the matrix

$ 
\textbf{D} = 
\begin{bmatrix}
5 & 0 & 0 \\
0 & -1 & 0 \\
0 & 0 & 3
\end{bmatrix}
$

#### Example 2

Show that the following matrix is not diagonalisable,

$ 
\textbf{A} = 
\begin{bmatrix}
3 & 0 & 0 \\
0 & 2 & 0 \\
0 & 1 & 2
\end{bmatrix}
$

**Solution:**

Start by evaluating the eigenvalues and eigenvectors

In [10]:
import numpy as np
import scipy as sp

A = np.array([[3, 0, 0], [0, 2, 0], [0, 1, 2]])

w, v= sp.linalg.eig(A)

print('List of Eigenvalues', w)
print('\nList of Eigenvectors\n', v.transpose())

List of Eigenvalues [3.+0.j 2.+0.j 2.+0.j]

List of Eigenvectors
 [[ 1.0000000e+00  0.0000000e+00  0.0000000e+00]
 [ 0.0000000e+00  0.0000000e+00  1.0000000e+00]
 [ 0.0000000e+00  4.4408921e-16 -1.0000000e+00]]


The matrix has eigenalues and eigenvectors

$ \lambda_1 = 3 $ has eigenvector
$ 
\begin{bmatrix}
1 \\
0 \\
0
\end{bmatrix}
$

$ \lambda_2 = 2 $ has eigenvectors
$ 
\begin{bmatrix}
0 \\
0 \\
1
\end{bmatrix}
$
and
$ 
\begin{bmatrix}
0 \\
0 \\
-1
\end{bmatrix}
$
which are not linearly independent.

$ \textbf{A} $ is a $ 3 \times 3 $ matrix whose eigenspace have only 2 basis vectors. So $ \textbf{A} $ is not diagonalisable.

## Powers of Matrix

The most efficient way to compute the power of a matrix.

If $ K $ is a positive integer, $ \lambda $ is an eigenvalue of a matrix $ \textbf{A} $, and $ \textbf{x} $ is a corresponding eigenvector, then $ \lambda^{K} $ is an eigenvalue of $ \textbf{A}^{K} $ and $ \textbf{x} $ is a corresponding eigenvector.

$ \textbf{P}^{-1} \textbf{A} \textbf{P} = $
$ 
\begin{bmatrix}
\lambda_1 & 0 & \cdots & 0 \\
0 & \lambda_2 & \cdots & 0 \\
\vdots \\
0 & 0 & \cdots & \lambda_n
\end{bmatrix}
= \textbf{D}
$

$ (\textbf{P}^{-1} \textbf{A} \textbf{P})^{K} = $
$ 
\begin{bmatrix}
\lambda_1^{K} & 0 & \cdots & 0 \\
0 & \lambda_2^{K} & \cdots & 0 \\
\vdots \\
0 & 0 & \cdots & \lambda_n^{K}
\end{bmatrix}
= \textbf{D}^{K}
$

**Proof:**

$ (\textbf{P}^{-1} \textbf{A} \textbf{P})^{K} = \textbf{P}^{-1} \textbf{A} \textbf{P} \textbf{P}^{-1} \textbf{A} \textbf{P} \cdots \textbf{P}^{-1} \textbf{A} \textbf{P} = \textbf{D}^{K}$

$ \Rightarrow (\textbf{P}^{-1} \textbf{A} \textbf{P})^{K} = \textbf{P}^{-1} \textbf{A} \cancel{\textbf{P}} \cancel{\textbf{P}^{-1}} \textbf{A} \cancel{\textbf{P}} \cdots \cancel{\textbf{P}^{-1}} \textbf{A} \textbf{P} \textbf{D}^{K} = \textbf{D}^{K} $

$ \Rightarrow (\textbf{P}^{-1} \textbf{A} \textbf{P})^{K} = \textbf{P}^{-1} \textbf{A}^{K} \textbf{P} = \textbf{D}^{K} $

$ \Rightarrow \textbf{A}^{K}  = \textbf{P} \textbf{D}^{K} \textbf{P}^{-1} $

Now $ \textbf{P} $ is just the matrix of eigenvectors of $ \textbf{A} $ and $ \textbf{D}^{K} $ is simply the $ K^{th} $ power of the diagonal elements of the matrix $ \textbf{D} $. Therefore calculating the power of the matrix $ \textbf{A} $ simply depends on evaluating its eigenvalues and eigenvectors.

### Example 3

Find $ \textbf{A}^{12} $ where 
$
\textbf{A} = 
\begin{bmatrix}
-1 & 1 & 0 \\
0 & 3 & 0 \\
4 & -2 & 5
\end{bmatrix}
$

**Solution:**

Find the eigenvalues and eigenvectors of $ \textbf{A} $

In [11]:
import numpy as np
import scipy as sp

A = np.array([[-1, 1, 0], [0, 3, 0], [4, -2, 5]])

w, v= sp.linalg.eig(A)

print('List of Eigenvalues', w)
print('\nList of Eigenvectors\n', v.transpose())

List of Eigenvalues [ 5.+0.j -1.+0.j  3.+0.j]

List of Eigenvectors
 [[ 0.          0.          1.        ]
 [ 0.83205029  0.         -0.5547002 ]
 [ 0.21821789  0.87287156  0.43643578]]


Therefore, the eigenvaules and eigenvectors are

$ \lambda_1 = 5.0 \ \ $ with eigenvector
$
\begin{bmatrix}
0 \\
0 \\
1 
\end{bmatrix}
$
,

$ \lambda_1 = -1.0 \ \ $ with eigenvector
$
\begin{bmatrix}
0.832 \\
0 \\
-0.555 
\end{bmatrix}
$
,

$ \lambda_1 = 3.0 \ \ $ with eigenvector
$
\begin{bmatrix}
0.218 \\
0.873 \\
0.436 
\end{bmatrix}
$

Therefore we can write

$ 
\textbf{D} =
\begin{bmatrix}
5 & 0 & 0 \\
0 & -1 & 0 \\
0 & 0 & 3
\end{bmatrix}
$

and 

$ 
\textbf{P} =
\begin{bmatrix}
0 & 0.832 & 0.218 \\
0 & 0 & 0.873 \\
1 & -0.555 & 0.436
\end{bmatrix}
$

Now applying $ \textbf{A}^{12}  = \textbf{P} \textbf{D}^{12} \textbf{P}^{-1} $

$ \textbf{D}^{12} $ is just the $ 12^{th} $ power of the diagonal elements of $ \textbf{D} $

In [15]:
import numpy as np
import scipy as sp

D_12 = np.array([[5**12, 0, 0], [0, (-1)**12, 0], [0, 0, 3**12]])

P = np.array([[0, 0.832, 0.218], [0, 0, 0.873], [1, -0.555, 0.436]])

P_inv = sp.linalg.inv(P)

print(P.dot(D_12).dot(P_inv))

[[ 1.00000000e+00  1.32707812e+05  0.00000000e+00]
 [ 0.00000000e+00  5.31441000e+05  0.00000000e+00]
 [ 1.62858229e+08 -1.62332988e+08  2.44140625e+08]]


Therefore 

$ 
\textbf{A}^{12} = 
\begin{bmatrix}
1 & 132,860 & 0 \\
0 & 531,441 & 0 \\
162,760,416 & -162,494,696 & 244,140,625
\end{bmatrix}
$

Let's check using SciPy

In [2]:
import numpy as np
import scipy as sp

A = np.array([[-1, 1, 0], [0, 3, 0], [4, -2, 5]])

print(A.dot(A).dot(A).dot(A).dot(A).dot(A).dot(A).dot(A).dot(A).dot(A).dot(A).dot(A))

[[         1     132860          0]
 [         0     531441          0]
 [ 162760416 -162494696  244140625]]


Which confirms our calculation.